internal/serialization/serialization.go

```diff
@@ -1,87 +1,89 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package serialization
 
 import (
 	"fmt"
 	"reflect"
 	"strconv"
 
-	"github.com/hazelcast/hazelcast-go-client/core"
-	"github.com/hazelcast/hazelcast-go-client/serialization"
+	"github.com/hazelcast/hazelcast-go-client/internal/hzerror"
+	pubserialization "github.com/hazelcast/hazelcast-go-client/serialization"
 )
 
 // Service serializes user objects to Data and back to Object.
 // Data is the internal representation of binary Data in Hazelcast.
 type Service struct {
-	serializationConfig *serialization.Config
-	registry            map[int32]serialization.Serializer
+	serializationConfig *pubserialization.Config
+	registry            map[int32]pubserialization.Serializer
 	nameToID            map[string]int32
 }
 
-func NewService(serializationConfig *serialization.Config) (*Service, error) {
+func NewService(serializationConfig *pubserialization.Config) (*Service, error) {
 	v1 := Service{serializationConfig: serializationConfig, nameToID: make(map[string]int32),
-		registry: make(map[int32]serialization.Serializer)}
+		registry: make(map[int32]pubserialization.Serializer)}
 	err := v1.registerDefaultSerializers()
 	if err != nil {
 		return nil, err
 	}
-	v1.registerCustomSerializers(serializationConfig.CustomSerializers())
-	v1.registerGlobalSerializer(serializationConfig.GlobalSerializer())
+	v1.registerCustomSerializers(serializationConfig.CustomSerializers)
+	v1.registerGlobalSerializer(serializationConfig.GlobalSerializer)
 	return &v1, nil
 }
 
 // ToData serializes an object to a Data.
 // It can safely be called with a Data. In that case, that instance is returned.
 // If it is called with nil, nil is returned.
-func (s *Service) ToData(object interface{}) (serialization.Data, error) {
-	if _, ok := object.(*Data); ok {
-		return object.(*Data), nil
+func (s *Service) ToData(object interface{}) (pubserialization.Data, error) {
+	if _, ok := object.(*SerializationData); ok {
+		return object.(*SerializationData), nil
 	}
-	dataOutput := NewPositionalObjectDataOutput(1, s, s.serializationConfig.IsBigEndian())
+	dataOutput := NewPositionalObjectDataOutput(1, s, s.serializationConfig.BigEndian)
 	serializer, err := s.FindSerializerFor(object)
 	if err != nil {
 		return nil, err
 	}
 	dataOutput.WriteInt32(0) // partition
 	dataOutput.WriteInt32(serializer.ID())
 	err = serializer.Write(dataOutput, object)
-	return &Data{dataOutput.buffer}, err
+	return &SerializationData{dataOutput.buffer[:dataOutput.position]}, err
```

Comment:
nit: access to private fields on `dataOutput` seems to be fragile. It makes sense to wrap it in a method on `ObjectDataOutput`:
```go
func (o *ObjectDataOutput) AsBufferSlice() []byte { // exact name doesn't matter
	return o.buffer[:o.position]
}
```